src/kernel/thread.c

```diff
@@ -627,7 +629,10 @@ void chargeBudget(ticks_t consumed, bool_t canTimeoutFault)
 
 void endTimeslice(bool_t can_timeout_fault)
 {
-    if (can_timeout_fault && !isRoundRobin(NODE_STATE(ksCurSC)) && validTimeoutHandler(NODE_STATE(ksCurThread))) {
+    bool_t round_robin = isRoundRobin(NODE_STATE(ksCurSC));
+    bool_t valid = validTimeoutHandler(NODE_STATE(ksCurThread));
+
+    if (can_timeout_fault && !round_robin && valid) {
```

Comment:
This is the bit that does change behaviour slightly, because `validTimeoutHandler` now must be executed no matter if `isRoundRobin` returns true or false. The new version has 3 `ldr` instructions that are executed unconditionally, and then a `cmp` and `beq` later. The old version has one `ldr`, a branch, then the other two `ldr` (I assume `validTimeoutHandler`), and then another comparison.

Reply:
Not sure if that comes out as slower or faster in the end, tbh.

Reply:
That would be true if `validTimeoutHandler` and `isRoundRobin` would be normal functions that can modify any memory. But in this case, both are inline and the compiler has complete view of all function bodies anyway, because seL4 concatenates all files, so it is free to optimise those calls away in case can_timeout_fault is false.

That is, it could have generated the old assembly if it wanted, but the changes were large enough to lead to different code generation.

Reply:
True, since they are inline and obviously read-only the compiler should be able to rearrange if it wants to.